crux-mir/src/Mir/Concurrency.hs

```diff
@@ -0,0 +1,157 @@
+       pure $! ThreadYield SimpleYield refs isRO
+  | otherwise = Nothing
+
+mirLock :: C.IsSymInterface sym => ExplorePrimitiveMatcher p sym MIR
+mirLock _ nm ctx cf _
+  | matchGeneric "core::crucible::concurrency::mutex_lock" nm
+  , Ctx.Empty Ctx.:> MirReferenceRepr t <- ctx
+  = do arg <- retrieveTypedArg ctx cf (MirReferenceRepr t) 0
+       let refs = mirRefName arg
+       case refs of
+         [ref] ->
+           pure $! ThreadYield (Acquire ref) refs False
+         _ -> error "TODO: Muxed mutex lock"
+  | matchGeneric "core::crucible::concurrency::mutex_unlock" nm
+  , Ctx.Empty Ctx.:> MirReferenceRepr t <- ctx
+  = do arg <- retrieveTypedArg ctx cf (MirReferenceRepr t) 0
+       let refs = mirRefName arg
+       case refs of
+         [ref] ->
+           pure $! ThreadYield (Release ref) refs False
+         _ -> error "TODO: Muxed mutex unlock"
+  | otherwise = Nothing
+
+mutexName :: BV.BV 32 -> Text.Text
+mutexName bv = Text.pack ("resource-"++ show bv)
+
+mirSpawn :: C.IsSymInterface sym => ExplorePrimitiveMatcher p sym MIR
+mirSpawn _ nm ctx cf@C.CallFrame { C._frameCFG = cfg } _
+  | matchGeneric "core::crucible::concurrency::spawn_internal" nm =
+  -- We use this as both the spawned thread and the function that returns the
+  -- spawned thread's ID. See the note in our libstd/thread/mod.rs
+    do C.Some ty <- retrieveType ctx 0
+       arg       <- retrieveTypedArg ctx cf ty 0
+       let hdl    = C.cfgHandle cfg
+           tpr    = C.BVRepr (W4.knownNat @32)
+           mkRet sym v = W4.bvLit sym W4.knownRepr (BV.mkBV BV.knownNat (fromIntegral v))
+       let check = ( W4.testEquality ctx (Ctx.Empty Ctx.:> ty)
+                   , W4.testEquality (C.handleArgTypes hdl) ctx
+                   , W4.testEquality (C.handleReturnType hdl) tpr
+                   )
+       case check of
+         (Just Refl, Just Refl, Just Refl) ->
+           return $ ThreadCreate ty tpr tpr hdl arg mkRet
+         _ ->
+           error $ "crucible_spawn_internal: unexpected signature "
+                ++ show (ctx, C.handleArgTypes hdl, C.handleReturnType hdl)
+  | otherwise = Nothing
+
+mirExit :: C.IsSymInterface sym => ExplorePrimitiveMatcher p sym MIR
+mirExit _ nm ctx cf _
+  | matchGeneric "core::crucible::concurrency::spawn::thread_exit" nm =
+    do (C.Some idx) <- Ctx.intIndex 0 (Ctx.size ctx)
+       return $ ThreadFinish $ C.Some (C.regVal' (cf ^. C.frameRegs) (C.Reg idx))
+  | otherwise
+  = Nothing
+
+mirRefName :: C.IsSymInterface sym => MirReferenceMux sym tp -> [Text.Text]
+mirRefName t =
+  case t of
+    MirReferenceMux tree ->
+       let refs = fst <$> viewFancyMuxTree tree
+           extr r =
+             case r of
+               MirReference root p ->
+                 Text.pack (show root ++ mirPathName p)
+               -- N.B.: I don't think we should ever hit this:
+               MirReference_Integer t rv ->
+                 error "Unexpected MirReference_Integer in mirRefName"
```

Comment:
Looking at how this gets called, I agree that under the current implementation this case will never be hit.  But that relies on the fact that `std::sys::crux::Mutex` has nonzero size, which IMO is a bit unintuitive.  Would it be easy enough to just add support for this case?  `rv` is just a symbolic `usize`, same as in the `Index_RefPath` case.

Reply:
Hm I think the connection between the size of `Mutex` and this constructor is a bit subtle. However, I suppose that since this denotes an integer interpreted as a pointer, the integer value itself should work fine as a name (i.e., address). 

If the pointer can never be dereferenced as indicated by the comment on the constructor, then I guess there will be no observable effects related to the resource. Is that right? Or is just that reads and writes will have to live in `unsafe` code? Is there a requirement that the result of the cast never aliases a value that would be denoted by a `MirReference`?

Reply:
On second thought, you should probably continue to forbid this, but maybe with a more informative comment (either here or on the definition of `Mutex`).

You could indeed use the integer/address as the name.  But if `Mutex` ever became zero-sized, you would run into the separate issue that multiple `Mutex`es that are conceptually distinct might have the same address.  You can see this in action with this code:
```Rust
    let v = vec![(), (), ()];
    println!("{:p} {:p} {:p}", &v[0], &v[1], &v[2]);
```
which prints `0x1 0x1 0x1`.  Since `()` is zero-sized (so dereferencing `&()` never accesses any memory), `Vec::index` is allowed to cast any non-null, well-aligned integer into a `&()` reference.  It happens to always choose the smallest legal value, which for `()` is `1`.  If `Mutex` was zero-sized, it would do the same thing there, so you could end up with a vector of mutexes where `ms[0]` and `ms[1]` actually have the same address (`MirReference_Integer 1`) internally.

Reply:
Thanks, I'll note that